src/dark_mind.c

```diff
@@ -0,0 +1,2218 @@
+        sub_08103A00(r4);
+        r5->unkDA = 3;
+    }
+    else if ((r3 -= sp04) < 0)
+    {
+        sub_08103A00(r4);
+        r5->unkDA = 5;
+    }
+    else if ((r3 -= ip) < 0)
+    {
+        sub_08103FC0(r4);
+    }
+    else if ((r3 -= r7) < 0)
+    {
+        sub_08103FC0(r4);
+        r5->unkDA = 7;
+        r5->unkD8 -= 0x10;
+    }
+    else if ((r3 -= sp00) < 0)
+    {
+        sub_08109B64(r4);
+    }
+    else if ((r3 -= sp08) < 0)
+    {
+        sub_08109B64(r4);
+        r5->unkDA = 9;
+    }
+    else if ((r3 -= sl) < 0)
+    {
+        sub_081044D0(r4);
+    }
+    else if ((r3 -= sp0C) < 0)
+    {
+        sub_08109B98(r4);
+    }
+    else if ((r3 -= sp10) < 0)
+    {
+        sub_08109B98(r4);
+        r5->unkDA = 11;
+        r5->unkD4 = 4;
+    }
+    else
+    {
+        r4->unk0.unk85 = 1;
+        r5->unkDA = 0;
+        sub_081030A8(r4);
+    }
+}
+
+void sub_081007A8(struct DarkMind *r3)
+{
+    switch (r3->unkD5)
+    {
+    default:
+    case 0 ... 2:
+        if (Rand16() & 1)
+            sub_08101968(r3);
+        else
+            sub_08100538(r3);
+        break;
+    case 3:
+        if (Rand16() & 7)
+            sub_08100538(r3);
+        else
+        {
+            r3->unkD5 = 0;
+            sub_08101630(r3);
+        }
+        break;
+    case 5:
+        if (Rand16() & 1)
+            sub_08101968(r3);
+        else
+            sub_08101630(r3);
+        break;
+    }
+}
+
+void sub_08100858(struct DarkMind *r3)
+{
+    u32 r5 = 1;
+    struct DarkMind *r4 = r3;
+    u16 r2;
+#ifdef NONMATCHING
+    s32 r0;
+#else
+    register s32 r0 asm("r0");
+#endif
+
+    switch (r3->unkD5)
+    {
+    default:
+    case 0:
+    case 1:
+        if (Rand16() & 1)
+        {
+            if (Rand16() & 1)
+                sub_08101968(r3);
+            else
+                sub_08101A40(r3);
+        }
+        else
+            sub_08100538(r3);
+        break;
+    case 2:
+    case 3:
+        r2 = Rand16() & 7;
+        if (r4->unkD5 == 2)
+        {
+            r5 = 3;
+            r0 = 2;
+        }
+        else
+            r0 = 5;
+        if (r2 < r5)
+        {
+            r3->unk0.unk85 = 0;
+            sub_081030A8(r3);
+        }
+        else if (r2 < (r0 = r5 + r0))
+            sub_08100538(r3);
+        else if (r2 < ++r0)
+        {
+            if (r4->unkD5 == 2)
+                sub_08101968(r3);
+            else
+                sub_08101A40(r3);
+        }
+        else
+            sub_08101630(r3);
+        break;
+    case 4:
+    case 5:
+        if (Rand16() & 1)
+        {
+            if (r4->unkD5 == 4)
+                sub_08101968(r3);
+            else
+                sub_08101A40(r3);
+        }
+        else if (Rand16() & 1)
+        {
+            if (r4->unkD5 == 4)
+                sub_08101A40(r3);
+            else
+                sub_08101968(r3);
+        }
+        else
+            sub_08101630(r3);
+        break;
+    }
+}
+
+#define Macro_08100BD0() \
```

Comment:
I've seen this pattern in some objects, maybe we can use this macro in other files?